src/server.rs

```diff
@@ -376,11 +404,41 @@ async fn index(
     middleware_router: web::Data<Arc<MiddlewareRouter>>,
     global_request_headers: web::Data<Arc<DashMap<String, String>>>,
     global_response_headers: web::Data<Arc<DashMap<String, String>>>,
+    calls_count: web::Data<Arc<DashMap<String, Vec<u64>>>>,
     body: Bytes,
     req: HttpRequest,
 ) -> impl Responder {
     let mut request = Request::from_actix_request(&req, body, &global_request_headers);
 
+    // Update calls count
+    let limit_key = format!(
+        "{}::{}::{}",
+        req.uri().path(),
+        req.method(),
+        req.peer_addr().map(|val| val.ip().to_string()).unwrap()
+    );
+    let current_timestamp = SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .unwrap()
+        .as_secs();
+    if let Some(mut timestamps) = calls_count.get_mut(&limit_key) {
+        let mut valid_timestamps: Vec<u64> = timestamps.iter().copied().collect();
+        valid_timestamps.push(current_timestamp);
+        *timestamps = valid_timestamps;
+    } else {
+        calls_count.insert(limit_key.to_owned(), vec![current_timestamp]);
+    }
+    let calls_count_header = calls_count
+        .get(&limit_key)
+        .unwrap()
+        .iter()
+        .map(ToString::to_string)
+        .collect::<Vec<String>>()
+        .join(",");
+    request
+        .headers
+        .insert(String::from("x-robyncc"), calls_count_header);
```

Comment:
Can you please explain the logic here?

Reply:
Sure. The rate limiting is done via a sliding window, however, in this part of the program, we don't know where the window starts. In order to pass it forward into the python part I'm embedding it in a custom header that I can later read and check how many calls fall in the allowed window.

Reply:
Thank you, makes sense 😄